src/planet-a/SubdivTriangleMesh.cpp

```diff
@@ -0,0 +1,170 @@
+    // worked out here: https://www.desmos.com/calculator/ffd8lraosl
+    unsigned int const sharedMax
+            = ceil( ( 3.0f*C + sqrt(6.0f)*sqrt(C) ) * pow(2.0f, L-1) - C + 1 );
+
+    unsigned int const fanMax
+            = ChunkedTriangleMeshInfo::smc_minFansVsLevel[subdivLevels] * 2;
+
+    return ChunkedTriangleMeshInfo(
+                chunkMax, subdivLevels, sharedMax, fanMax);
+}
+
+ChunkId ChunkedTriangleMeshInfo::chunk_create(
+        SubdivTriangleSkeleton& rSkel,
+        SkTriId skTri,
+        ArrayView_t<SkVrtxId> const edgeRte,
+        ArrayView_t<SkVrtxId> const edgeBtm,
+        ArrayView_t<SkVrtxId> const edgeLft)
+{
+    if (   edgeRte.size() != m_chunkWidth - 1
+        || edgeBtm.size() != m_chunkWidth - 1
+        || edgeLft.size() != m_chunkWidth - 1)
+    {
+        throw std::runtime_error("Incorrect edge vertex count");
+    }
+
+    // Create a new chunk ID
+    ChunkId const chunkId = m_chunkIds.create();
+    Chunk &rChunk = m_chunkData[size_t(chunkId)];
+
+    rChunk.m_skeletonTri = rSkel.tri_store(skTri);
+
+    SkeletonTriangle const &tri = rSkel.tri_at(skTri);
+
+    ArrayView_t<SharedVrtxStorage_t> const sharedSpace = chunk_shared_mutable(chunkId);
+
+    std::array const edges = {edgeRte, edgeBtm, edgeLft};
+
+    for (int i = 0; i < 3; i ++)
+    {
+        size_t const cornerOffset = m_chunkWidth * i;
+        size_t const edgeOffset = m_chunkWidth * i + 1;
+
+        ArrayView_t<SkVrtxId> const edge = edges[i];
+        {
+            SharedVrtxId const sharedId = shared_get_or_create(
+                    tri.m_vertices[i], rSkel);
+            sharedSpace[cornerOffset] = shared_store(sharedId);
+        }
+        for (unsigned int j = 0; j < m_chunkWidth - 1; j ++)
+        {
+            SharedVrtxId const sharedId = shared_get_or_create(edge[j], rSkel);
+            sharedSpace[edgeOffset + j] = shared_store(sharedId);
+        }
+    }
+
+    return chunkId;
+}
+
+//-----------------------------------------------------------------------------
+
+ChunkVrtxSubdivLUT::ChunkVrtxSubdivLUT(uint8_t subdivLevel)
+{
+    m_edgeVrtxCount = 1u << subdivLevel;
+    m_fillVrtxCount = (m_edgeVrtxCount-2) * (m_edgeVrtxCount-1) / 2;
+
+    m_data.reserve(m_fillVrtxCount);
+
+    // Calculate LUT, this fills m_data
+    fill_tri_recurse({0,               0              },
+                     {0,               m_edgeVrtxCount},
+                     {m_edgeVrtxCount, m_edgeVrtxCount},
+                     subdivLevel);
+
+    // Future optimization:
+    // m_data can be sorted in a way that slightly improves cache locality
+    // by accessing to fill vertices in a more sequential order.
+}
+
+void ChunkVrtxSubdivLUT::subdiv_line_recurse(
+        Vector2us a, Vector2us b, uint8_t level)
+{
+    Vector2us const mid = (a + b) / 2;
+
+    ChunkLocalSharedId out = ChunkLocalSharedId(xy_to_triangular(mid.x() - 1, mid.y() - 2));
+
+    m_data.emplace_back(ToSubdiv{id_at(a), id_at(b), out});
```

Comment:
weird to emplace_back a function call result?

push_back(), or alternatively some way to make the constructor of m_data::value_type call ToSubdiv or what have you.

Reply:
ToSubdiv{..} creates a POD struct rvalue? I intending this to be calling ToSubdiv's default move construct, and optimizing down to in-place construction.